flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FileMergingCheckpointStateOutputStream.java

```diff
@@ -0,0 +1,255 @@
+    int bufferPos;
+
+    public FileMergingCheckpointStateOutputStream(
+            long checkpointID,
+            int bufferSize,
+            FileMergingSnapshotManagerProxy fileMergingSnapshotManagerProxy) {
+        this.fileMergingSnapshotManagerProxy = fileMergingSnapshotManagerProxy;
+        this.writeBuffer = new byte[bufferSize];
+    }
+
+    /** Assign a physical file to this stream and initialize the outputStream. */
+    private void initializeOutputStream() throws IOException {
+        Tuple2<FSDataOutputStream, Path> streamAndPath =
+                fileMergingSnapshotManagerProxy.providePhysicalFile();
+        outputStream = streamAndPath.f0;
+        startPos = outputStream.getPos();
+        filePath = streamAndPath.f1;
+    }
+
+    @Override
+    public long getPos() throws IOException {
+        // The starting position is not determined until a physical file has been assigned, so
+        // we return the relative value to the starting position in this method
+        return bufferPos + curPosRelative;
+    }
+
+    @Override
+    public void write(int b) throws IOException {
+        if (bufferPos >= writeBuffer.length) {
+            flushToFile();
+        }
+        writeBuffer[bufferPos++] = (byte) b;
+    }
+
+    @Override
+    public void write(byte[] b, int off, int len) throws IOException {
+        if (len < writeBuffer.length) {
+            // copy it into our write buffer first
+            final int remaining = writeBuffer.length - bufferPos;
+            if (len > remaining) {
+                // copy as much as fits
+                System.arraycopy(b, off, writeBuffer, bufferPos, remaining);
+                off += remaining;
+                len -= remaining;
+                bufferPos += remaining;
+
+                // flushToFile the write buffer to make it clear again
+                flushToFile();
+            }
+
+            // copy what is in the buffer
+            System.arraycopy(b, off, writeBuffer, bufferPos, len);
+            bufferPos += len;
+        } else {
+            // flushToFile the current buffer, outputStream is initialized if it is null
+            flushToFile();
+            // write the bytes directly
+            outputStream.write(b, off, len);
+            curPosRelative += len;
+        }
+    }
+
+    @Override
+    public void flush() throws IOException {
+        if (outputStream != null) {
+            flushToFile();
+        }
+    }
+
+    @Override
+    public void sync() throws IOException {
+        if (outputStream != null) {
+            outputStream.sync();
+        }
+    }
+
+    @Nullable
+    @Override
+    public SegmentFileStateHandle closeAndGetHandle() throws IOException {
+        // check if there was nothing ever written
+        if (outputStream == null && bufferPos == 0) {
+            return null;
+        }
+
+        synchronized (this) {
+            if (!closed) {
+                try {
+                    flushToFile();
```

Comment:
This class behaves similarly to ```FsCheckpointStateOutputStream```, which never calls ```outputStream.flush()```, but relies on ```outputStream.close()``` to ensure the persistency of data.

So, for ```flush()``` and ```sync()```, maybe we do not need to call ```outputStream.close()```.

Whereas we do need to think about ```closeAndGetHandle()```. I think we should call ```outputStream.sync()```  before returning the ```SegmentFileStateHandle``` instance, to make sure the file segment is visible. WDYT?

Reply:
If so, we'd better note this down since it violates the behavior of the base interface.

And for calling ```outputStream.sync()``` in ```closeAndGetHandle()```, I think it has been already covered by ```returnPhysicalFileForNextReuse```, WDYT?

Reply:
As for `closeAndGetHandle()`: Yeah, you are right. No need for an extra `outputStream.sync()` as the `returnPhysicalFileForNextReuse ` can handle it. I'd forgotten that.